### PR TITLE
Adaptive Load FakeStepController doom update

### PR DESCRIPTION
**Pull request by eric846**
Updates FakeStepController:

- Set doomed state when `FakeStepController::UpdateAndRecompute()` receives a `BenchmarkResult` with a negative metric score. Previously `BenchmarkResult` could contain an error status, and we used the error status to trigger simulated doom. Now errors do not get passed to  `StepController::UpdateAndRecompute()`, so we need a new mechanism.
- Adds the ability to simulate an `InputVariableSetter` failure in `GetCurrentCommandLineOptions()`. The artificial failure is specified in the config proto.
- Adds a countdown mechanism so that `GetCurrentCommandLineOptions()` returns successfully until `UpdateAndRecompute()` has been called a configured number of times, then starts returning the simulated input variable setter failure mentioned above. This simple autonomous behavior is needed to test error handling in the testing stage without triggering an early exit in the adjusting stage; the adaptive load controller can't be paused in the middle of the test to reconfigure the `FakeStepController` to change its behavior.

The behavior of `FakeStepController` after an `UpdateAndRecompute()` with a `BenchmarkResult` containing the given metric scores:
- Any positive score: converged
- All scores zero: neither converged nor doomed
- Any negative score: doomed

Part 3 of splitting PR #483.

### CODE REVIEW

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto [hunk 1]
```diff
@@ -5,12 +5,23 @@ package nighthawk.adaptive_load;
...
+// from input value setting every time, or return an error after some number of UpdateAndRecompute()
+// iterations.
 message FakeStepControllerConfig {
-  // RPS that should always be returned. Optional, default 0.
+  // RPS that should always be returned, except when artificial errors are configured. Optional,
+  // default 0.
   int32 fixed_rps_value = 1;
   // Artificial error that the plugin factory should return during validation. Optional.
   google.rpc.Status artificial_validation_failure = 2;
+  // Artificial error that should be returned from GetCurrentCommandLineOptions(). Optional.
```
Comment (dubious90) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto:19]: Can we enhance this comment to indicate the relation to the countdown?

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto [hunk 2]
```diff
@@ -5,12 +5,23 @@ package nighthawk.adaptive_load;
...
+  google.rpc.Status artificial_input_setting_failure = 3;
```
Comment (dubious90) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto:20]: I think I'm having trouble understanding the full use case here.  If this is 3, then we are going to succeed 3 times, then fail on the 4th attempt. Makes sense, but I'm not sure I understand why that's useful.  What is the test that you're supporting by creating it in this way?

**Comment by dubious90** [comment]
@mum4k Can you give this a review and assign back to me when done?

**Comment by mum4k** [comment]
@eric846 please address / respond to the comments from @dubious90 as that can help me by providing additional context for the review.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.cc [hunk 3]
```diff
@@ -15,9 +15,10 @@ absl::Status StatusFromProtoRpcStatus(const google::rpc::Status& status_proto) {
 } // namespace
 
 FakeStepController::FakeStepController(
-    const nighthawk::adaptive_load::FakeStepControllerConfig& config,
+    nighthawk::adaptive_load::FakeStepControllerConfig config,
     nighthawk::client::CommandLineOptions command_line_options_template)
-    : is_converged_{false}, is_doomed_{false}, fixed_rps_value_{config.fixed_rps_value()},
+    : input_setting_failure_countdown_{config.artificial_input_setting_failure_countdown()},
+      config_{std::move(config)}, is_converged_{false}, is_doomed_{false},
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.cc:21]: Funny story, this was actually required by clang-tidy. If the function implementation is making a copy of a const reference parameter, clang-tidy tells you to change the parameter to be pass by value so it's obvious to the caller that a copy is happening. The argument gets copied at the call site into an unnamed temporary value. Then in the constructor we can actually move this temporary value into the field without incurring the cost of a second copy.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto (hunk 1 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto:19]: Done

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto (hunk 2 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.proto:20]: `GetCurrentCommandLine()` can return an error status that should be handled cleanly by the main controller loop. The controller calls `GetCurrentCommandLine()` repeatedly during the adjusting stage until the step controller says it's converged, and then in the testing stage we call `GetCurrentCommandLine()` one last time, reusing the last converged value. 

In order to test handling of these errors, we need the `FakeStepController` to return successfully from `GetCurrentCommandLine()` during the adjusting stage, but then start returning errors just in time for the testing stage. We don't have any way to update the `FakeStepController` during the run, so it has to somehow be programmed up front to behave differently at different times.

An alternative would be for magic values in `UpdateAndRecompute()` to trigger `GetCurrentCommandLine()` error behavior. We already use magic values to control convergence and doom. But there's only so much information we can encode in metric score `double`s without having it get out of hand. Currently the `UpdateAndRecompute()` behavior is: zero scores=non-converged non-doomed, any positive score=converged, any negative score=doomed.

This trick wouldn't be necessary if the step controller were aware of what stage it was operating in.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc [hunk 4]
```diff
@@ -163,5 +239,29 @@ TEST(MakeFakeStepControllerPluginConfigWithValidationError,
   EXPECT_EQ(plugin_or.status().message(), kValidationErrorMessage);
 }
 
+TEST(MakeFakeStepControllerPluginConfigWithInputSettingError,
+     ProducesFakeStepControllerPluginWithConfiguredErrorAndCountdown) {
+  const int kExpectedRpsValue = 123;
+  std::string kInputSettingErrorMessage = "artificial input setting error";
```
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:245]: const

**Review by oschaaf — APPROVED** [review]
Just a const nit, LGTM otherwise. Left a question that popped up over at https://github.com/envoyproxy/nighthawk/issues/466#issuecomment-682442611 as that's out of scope here.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.h [hunk 5]
```diff
@@ -22,7 +22,7 @@ class FakeStepController : public StepController {
    * @param config FakeStepControllerConfig proto for setting the fixed RPS value.
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.h:22]: Done

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.h [hunk 6]
```diff
@@ -100,4 +103,21 @@ MakeFakeStepControllerPluginConfig(int fixed_rps_value);
 envoy::config::core::v3::TypedExtensionConfig MakeFakeStepControllerPluginConfigWithValidationError(
     const absl::Status& artificial_validation_error);
 
+/**
+ * Creates a valid TypedExtensionConfig proto that activates a FakeStepController with a
+ * FakeInputVariableSetterConfig that returns an error from GetCurrentCommandLineOptions().
+ *
+ * @param fixed_rps_value Value for RPS to set in the FakeStepControllerConfig proto until the
+ * countdown reaches zero.
+ * @param artificial_input_setting_failure An error status.
```
[the remark this answers is not in the document]
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller.h:112]: Done

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc (hunk 4 above)
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:245]: Done

**Review by oschaaf — APPROVED** [review]
Thanks for resolving the const nit. LGTM for my part of the review.

**Pull request merged**